### PR TITLE
`azurerm_eventhub_namespace` / `azurerm_servicebus_namespace` - deprecate the `zone_redundant` field in v4.0

### PR DESCRIPTION
**Pull request by WodansSon**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description

The `zone_redundant` field, per the service team, is now going to be 100% controlled by the API. So exposing this as a field to the end user no longer makes sense.

## PR Checklist

- [X] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [X] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [X] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [X] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [X] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [X] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [X] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [X] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [X] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

### EventHub Test Results:
image

**Note:** The one failing test case is an environment issue not related to the code change in this PR.

### ServiceBus Test Results:
image

**NOTE:** The two failures are v4.0 related.

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_eventhub_namespace` / `azurerm_servicebus_namespace` - deprecate the `zone_redundant` field in v4.0

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/eventhub/eventhub_namespace_resource_test.go [hunk 1]
```diff
@@ -399,7 +400,7 @@ func TestAccEventHubNamespace_BasicWithSkuUpdate(t *testing.T) {
 			),
 		},
 		{
-			Config: r.premium(data),
+			Config: r.premium(data, features.FourPointOhBeta()),
```
Comment (stephybun) [inline internal/services/eventhub/eventhub_namespace_resource_test.go:403]: I appreciate the effort here to reduce the lines of code and repetition, but keeping the clean up effort of all the `FourPointOh`/`FourPointOhBeta` code in mind, this is more work to unpick than if the config is just duplicated.

We try and leave the test alone if possible and update the function returning the config to return the right config based on the flag, so in this case

```
func (EventHubNamespaceResource) premium(data acceptance.TestData) string {
	if !features.FourPointOhBeta() {
		return fmt.Sprintf(`
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-eh-%d"
  location = "%s"
}

resource "azurerm_eventhub_namespace" "test" {
  name                = "acctesteventhubnamespace-%d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  sku                 = "Premium"
  capacity            = "1"
  zone_redundant      = true
}
`, data.RandomInteger, data.Locations.Primary, data.RandomInteger)
	}
	
	return fmt.Sprintf(`
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "test" {
  name     = "acctestRG-eh-%d"
  location = "%s"
}

resource "azurerm_eventhub_namespace" "test" {
  name                = "acctesteventhubnamespace-%d"
  location            = azurerm_resource_group.test.location
  resource_group_name = azurerm_resource_group.test.name
  sku                 = "Premium"
  capacity            = "1"
}
`, data.RandomInteger, data.Locations.Primary, data.RandomInteger)
}
```

File: internal/services/servicebus/servicebus_namespace_resource_test.go [hunk 2]
```diff
@@ -151,6 +152,10 @@ func TestAccAzureRMServiceBusNamespace_premiumMessagingPartition(t *testing.T) {
 }
 
 func TestAccAzureRMServiceBusNamespace_zoneRedundant(t *testing.T) {
+	if features.FourPointOhBeta() {
+		t.Skipf("Skippped as 'zone_redundant' property is deprecated in 4.0")
```
Comment (stephybun) [inline internal/services/servicebus/servicebus_namespace_resource_test.go:156]:
```suggestion
		t.Skipf("Skipped as 'zone_redundant' property is deprecated in 4.0")
```

File: website/docs/d/eventhub_namespace.html.markdown [hunk 3]
```diff
@@ -44,6 +44,8 @@ output "eventhub_namespace_id" {
 
 * `zone_redundant` - Is this EventHub Namespace deployed across Availability Zones?
 
+~> **Note:** The `zone_redundant` field will be removed in v4.0 of the provider.
```
Comment (stephybun) [inline website/docs/d/eventhub_namespace.html.markdown:47]: Anyone using the field will be notified by the deprecation messages added to the field so no need to add this to the documentation as well
```suggestion
```

File: website/docs/d/servicebus_namespace.html.markdown [hunk 4]
```diff
@@ -41,6 +41,8 @@ output "location" {
 
 * `zone_redundant` - Whether or not this ServiceBus Namespace is zone redundant.
 
+~> **Note:** The `zone_redundant` field will be removed in v4.0 of the provider.
```
Comment (stephybun) [inline website/docs/d/servicebus_namespace.html.markdown:44]:
```suggestion
```

File: website/docs/r/eventhub_namespace.html.markdown [hunk 5]
```diff
@@ -55,7 +55,9 @@ The following arguments are supported:
 
 * `zone_redundant` - (Optional) Specifies if the EventHub Namespace should be Zone Redundant (created across Availability Zones). Changing this forces a new resource to be created. Defaults to `false`.
 
-~> **Note:** For eventhub premium namespace, `zone_redundant` is computed by api based on the availability zone feature in each region. User's input will be overridden. Please explicitly sets the property to `true` when creating the premium namespace in a region that supports availability zone since the default value is `false` in 3.0 provider.
+~> **Note:** For eventhub premium namespace, `zone_redundant` is computed by the api based on the availability zone feature in each region, user's input will be overridden. Please explicitly sets the property to `true` when creating the premium namespace in a region that supports availability zone since the default value is `false` in 3.0 provider.
+
+~> **Note:** The `zone_redundant` field will be removed in v4.0 of the provider.
```
Comment (stephybun) [inline website/docs/r/eventhub_namespace.html.markdown:60]:
```suggestion
```

File: website/docs/r/servicebus_namespace.html.markdown [hunk 6]
```diff
@@ -65,6 +65,8 @@ The following arguments are supported:
 
 * `zone_redundant` - (Optional) Whether or not this resource is zone redundant. `sku` needs to be `Premium`. Changing this forces a new resource to be created.
 
+~> **Note:** The `zone_redundant` field will be removed in v4.0 of the provider.
```
Comment (stephybun) [inline website/docs/r/servicebus_namespace.html.markdown:68]:
```suggestion
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Thanks for this @WodansSon. The breaking schema changes look good! The test needs some slight re-work though.

File: internal/services/eventhub/eventhub_namespace_resource_test.go (hunk 1 above)
Comment (WodansSon) [reply internal/services/eventhub/eventhub_namespace_resource_test.go]: Fixed.

File: internal/services/servicebus/servicebus_namespace_resource_test.go (hunk 2 above)
Comment (WodansSon) [reply internal/services/servicebus/servicebus_namespace_resource_test.go]: Fixed.

File: website/docs/d/eventhub_namespace.html.markdown (hunk 3 above)
Comment (WodansSon) [reply website/docs/d/eventhub_namespace.html.markdown]: Fixed.

File: website/docs/d/servicebus_namespace.html.markdown (hunk 4 above)
Comment (WodansSon) [reply website/docs/d/servicebus_namespace.html.markdown]: Fixed.

File: website/docs/r/eventhub_namespace.html.markdown (hunk 5 above)
Comment (WodansSon) [reply website/docs/r/eventhub_namespace.html.markdown]: Fixed.

File: website/docs/r/servicebus_namespace.html.markdown (hunk 6 above)
Comment (WodansSon) [reply website/docs/r/servicebus_namespace.html.markdown]: Fixed.

File: internal/services/eventhub/eventhub_namespace_data_source.go [hunk 7]
```diff
@@ -111,6 +107,15 @@ func EventHubNamespaceDataSource() *pluginsdk.Resource {
 			"tags": commonschema.TagsDataSource(),
 		},
 	}
+
+	if !features.FourPointOhBeta() {
+		resource.Schema["zone_redundant"] = &pluginsdk.Schema{
+			Type:     pluginsdk.TypeBool,
+			Computed: true,
+		}
```
Comment (katbyte) [inline internal/services/eventhub/eventhub_namespace_data_source.go:115]: we should still add a deprecation message here?

File: internal/services/servicebus/servicebus_namespace_data_source.go [hunk 8]
```diff
@@ -92,6 +88,15 @@ func dataSourceServiceBusNamespace() *pluginsdk.Resource {
 			"tags": tags.SchemaDataSource(),
 		},
 	}
+
+	if !features.FourPointOhBeta() {
+		resource.Schema["zone_redundant"] = &pluginsdk.Schema{
+			Type:     pluginsdk.TypeBool,
+			Computed: true,
+		}
```
Comment (katbyte) [inline internal/services/servicebus/servicebus_namespace_data_source.go:96]: and here

**Review by katbyte — CHANGES_REQUESTED** [review]
test failure:
```
------- Stdout: -------
=== RUN   TestAccAzureRMServiceBusNamespace_basicCapacity
=== PAUSE TestAccAzureRMServiceBusNamespace_basicCapacity
=== CONT  TestAccAzureRMServiceBusNamespace_basicCapacity
    testcase.go:113: Step 1/1, expected an error with pattern, no match on: Error running apply: exit status 1
        
        Error: service bus SKU "Basic" only supports `capacity` of 0
        
          with azurerm_servicebus_namespace.test,
          on terraform_plugin_test.tf line 30, in resource "azurerm_servicebus_namespace" "test":
          30: resource "azurerm_servicebus_namespace" "test" {
        
--- FAIL: TestAccAzureRMServiceBusNamespace_basicCapacity (42.32s)
FAIL

```

File: internal/services/eventhub/eventhub_namespace_data_source.go (hunk 7 above)
Comment (WodansSon) [reply internal/services/eventhub/eventhub_namespace_data_source.go:116]: Fixed.

File: internal/services/servicebus/servicebus_namespace_data_source.go (hunk 8 above)
Comment (WodansSon) [reply internal/services/servicebus/servicebus_namespace_data_source.go:97]: Fixed.

**Review by katbyte — APPROVED** [review]
LGTM ❄️

**Pull request merged**